src/main/java/org/mastodon/mamut/feature/branch/BranchNLeavesFeatureComputer.java

```diff
@@ -0,0 +1,55 @@
+package org.mastodon.mamut.feature.branch;
+
+import org.mastodon.collection.RefSet;
+import org.mastodon.graph.algorithm.LeafFinder;
+import org.mastodon.graph.algorithm.traversal.InverseDepthFirstIterator;
+import org.mastodon.mamut.feature.MamutFeatureComputer;
+import org.mastodon.mamut.model.branch.BranchLink;
+import org.mastodon.mamut.model.branch.BranchSpot;
+import org.mastodon.mamut.model.branch.ModelBranchGraph;
+import org.mastodon.properties.IntPropertyMap;
+import org.scijava.ItemIO;
+import org.scijava.plugin.Parameter;
+import org.scijava.plugin.Plugin;
+
+/**
+ * Computes {@link BranchNLeavesFeature}
+ */
+@Plugin( type = MamutFeatureComputer.class )
+public class BranchNLeavesFeatureComputer implements MamutFeatureComputer
+{
+	@Parameter
+	private ModelBranchGraph branchGraph;
+
+	@Parameter( type = ItemIO.OUTPUT )
+	private BranchNLeavesFeature output;
+
+	@Override
+	public void createOutput()
+	{
+		if ( null == output )
+			output = new BranchNLeavesFeature( new IntPropertyMap<>( branchGraph.vertices().getRefPool(), 0 ) );
+	}
+
+	@Override
+	public void run()
+	{
+		computeNumberOfSubtreeLeaves();
+	}
+
+	private void computeNumberOfSubtreeLeaves()
+	{
+		final RefSet< BranchSpot > leaves = LeafFinder.getLeaves( branchGraph );
+		for ( final BranchSpot leaf : leaves )
+		{
+			InverseDepthFirstIterator< BranchSpot, BranchLink > inverseDepthFirstIterator =
+					new InverseDepthFirstIterator<>( branchGraph );
+			inverseDepthFirstIterator.reset( leaf );
+			while ( inverseDepthFirstIterator.hasNext() )
+			{
+				BranchSpot branchSpot = inverseDepthFirstIterator.next();
+				output.map.set( branchSpot, output.map.get( branchSpot ) + 1 );
+			}
+		}
```

Comment:
The `DepthFirstSearch` class that is part of mastodon should be used to implement this method

```suggestion
	private void computeNumberOfSubtreeLeaves()
	{

		DepthFirstSearch< BranchSpot, BranchLink > search = new DepthFirstSearch<>( branchGraph, GraphSearch.SearchDirection.DIRECTED );
		search.setTraversalListener( new SearchListener< BranchSpot, BranchLink, DepthFirstSearch< BranchSpot, BranchLink > >() {
			@Override
			public void processVertexLate( BranchSpot vertex, DepthFirstSearch< BranchSpot, BranchLink > search )
			{
				boolean isLeaf = vertex.outgoingEdges().isEmpty();
				if ( isLeaf )
					output.map.set( vertex, 1 );
				else
				{
					BranchSpot ref = branchGraph.vertexRef();
					int n = 0;
					for ( BranchLink link : vertex.outgoingEdges() )
					{
						BranchSpot child = link.getTarget( ref );
						n += output.map.get( child );
					}
					output.map.set( vertex, n );
					branchGraph.releaseRef( ref );
				}
			}

			@Override
			public void processVertexEarly( BranchSpot vertex, DepthFirstSearch< BranchSpot, BranchLink > search )
			{

			}

			@Override
			public void processEdge( BranchLink edge, BranchSpot from, BranchSpot to, DepthFirstSearch< BranchSpot, BranchLink > search )
			{

			}

			@Override
			public void crossComponent( BranchSpot from, BranchSpot to, DepthFirstSearch< BranchSpot, BranchLink > search )
			{

			}
		} );

		for ( BranchSpot spot : branchGraph.vertices() )
		{
			boolean isRoot = spot.incomingEdges().isEmpty();
			if ( isRoot )
				search.start( spot );
		}
	}
```

Reply:
Thanks for introducing me to the respective classes. I took most parts of your code suggestion, but changed 

![grafik](https://user-images.githubusercontent.com/10515534/222222360-f1062497-c65d-4ec9-a9bb-0038bbe0ccec.png)

to 

![grafik](https://user-images.githubusercontent.com/10515534/222222438-08649050-2fdd-4211-b766-f057dfd7f56e.png)

since there is already the class RootFinder for finding roots.

I agree that it should be a lot faster, since lots of traversals from any leave to the roots are omitted and are replaced by just one DepthFirstSearch per root node. The already existing unit test do also pass with the suggested code.